gdpr/validate_block.go

```diff
@@ -0,0 +1,230 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package gdpr
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/golang/protobuf/proto"
+
+	"crypto/sha256"
+
+	"github.com/hyperledger/fabric-protos-go/common"
+	"github.com/hyperledger/fabric/core/ledger/kvledger/txmgmt/rwsetutil"
+	"github.com/hyperledger/fabric/protoutil"
+)
+
+var ErrVal = errors.New("gdpr: block validation failed")
+
+func getHash(preimage []byte) []byte {
+	h := sha256.New()
+	return h.Sum(preimage)
+
+}
+
+func validate(block *common.Block) (*common.Block, error) {
+	preImages := block.GetData().GetPreimageSpace()
+
+	m := map[string]struct{}{}
+
+	for _, im := range preImages {
+		temp := getHash(im)
+		m[(string(temp))] = struct{}{}
+	}
+
+	for _, envBytes := range block.Data.Data {
+		env, err := protoutil.GetEnvelopeFromBlock(envBytes)
+		if err != nil {
+			return nil, err
+		}
+
+		payload, err := protoutil.UnmarshalPayload(env.Payload)
+		if err != nil {
+			return nil, err
+		}
+
+		tx, err := protoutil.UnmarshalTransaction(payload.Data)
+		if err != nil {
+			return nil, err
+		}
+
+		_, respPayload, err := protoutil.GetPayloads(tx.Actions[0])
+		if err != nil {
+			return nil, err
+		}
+
+		txRWSet := &rwsetutil.TxRwSet{}
+
+		if err = txRWSet.FromProtoBytes(respPayload.Results); err != nil {
+			return nil, err
+		}
+
+		for _, nsRWSet := range txRWSet.NsRwSets {
+			for _, kvWrite := range nsRWSet.KvRwSet.Writes {
+				var a = kvWrite.GetValueHash()
+				if a != nil {
+					fmt.Printf("that's odd")
+				}
+				var b = getHash(kvWrite.GetValue())
+				if memberOf((string)(b), m) == false {
+					return nil, ErrVal
+				}
+
+			}
+
+		}
+	}
+
+	return getVanillaBlock(block)
+}
+
+func memberOf(a string, m map[string]struct{}) bool {
+	_, exists := m[a]
+	return exists
+
+}
+
+func getVanillaBlock(block *common.Block) (*common.Block, error) {
+	newBlock := proto.Clone(block).(*common.Block)
+	newBlock.Data.PreimageSpace = nil
+
+	clearKVWrites(newBlock) // TODO: remove when blocks are generated without KVWrite values.
```

Comment:
i don't see you error in case a pre-image isn't found, so why are you removing them if you intend to replace all of them anyway? Is it for the case you do not find a pre-image?

Reply:
Not sure I understand - I'm removing them to create a block in the old format, where there are no pre-images. Should I just keep this field as is when cloning?

Reply:
This function should assume that it accepts input that adheres to the new structure. You must not delete the KVWrites here because it will affect the block hash and as this is a utility function, you do not know who is going to consume it. 

All we need to do in the production code, is check hashes match, and then replace all hashes with pre-images.

Also, I think we should delete the pre-image space from the block we return since we move the pre-images to the KVWrites there is no justification for the pre-image space to exist in the return value any longer.

If by some mistake there is a peer that produces endorsements with KVWrites then there is nothing we should do about it in this function.

Furthermore, imagine that maybe we'd want to support a hybrid technique where some keys are sensitive and some are not, and the non-sensitive keys are not written to the pre-image space and are written using the legacy way.